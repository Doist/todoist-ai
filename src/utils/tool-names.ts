/**
 * Centralized tool names module
 *
 * This module provides a single source of truth for all tool names used throughout the codebase.
 * Each tool should import its own name from this module to avoid hardcoded strings.
 * This prevents outdated references when tool names change.
 */
export const ToolNames = {
    // Task management tools
    ADD_TASKS: 'add-tasks',
    COMPLETE_TASKS: 'complete-tasks',
    UPDATE_TASKS: 'update-tasks',
    FIND_TASKS: 'find-tasks',
    FIND_TASKS_BY_DATE: 'find-tasks-by-date',
    FIND_COMPLETED_TASKS: 'find-completed-tasks',

    // Project management tools
    ADD_PROJECTS: 'add-projects',
    UPDATE_PROJECTS: 'update-projects',
    FIND_PROJECTS: 'find-projects',

    // Section management tools
    ADD_SECTIONS: 'add-sections',
    UPDATE_SECTIONS: 'update-sections',
    FIND_SECTIONS: 'find-sections',

    // Comment management tools
    ADD_COMMENTS: 'add-comments',
    UPDATE_COMMENTS: 'update-comments',
    FIND_COMMENTS: 'find-comments',

    // Assignment and collaboration tools
    FIND_PROJECT_COLLABORATORS: 'find-project-collaborators',
    MANAGE_ASSIGNMENTS: 'manage-assignments',

    // General tools
    GET_OVERVIEW: 'get-overview',
    DELETE_OBJECT: 'delete-object',
    USER_INFO: 'user-info',
    QUICK_ADD_TASK: 'quick-add-task',
} as const

// Type for all tool names
export type ToolName = (typeof ToolNames)[keyof typeof ToolNames]
