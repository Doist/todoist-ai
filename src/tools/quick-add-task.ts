// Quick Add Task tool for Todoist MCP

import type { QuickAddTaskArgs, Task, TodoistApi } from '@doist/todoist-api-typescript'
import { z } from 'zod'
import { getToolOutput } from '../mcp-helpers.js'
import type { TodoistTool } from '../todoist-tool.js'
import { mapTask } from '../tool-helpers.js'
import { ToolNames } from '../utils/tool-names.js'

const QuickAddTaskSchema = z.object({
    text: z.string().min(1).describe('Write what the task is. This is required.'),
    note: z.string().optional().describe('Add a note to the task if you want.'),
    reminder: z.string().optional().describe('Set a reminder for the task if needed.'),
    autoReminder: z
        .boolean()
        .optional()
        .describe('If true, a default reminder is added when there is a due time.'),
    meta: z
        .boolean()
        .optional()
        .describe('If true, extra information about the task will be returned.'),
})

const ArgsSchema = {
    tasks: z.array(QuickAddTaskSchema).min(1).describe('List of tasks to quick add.'),
}

const quickAddTask = {
    name: ToolNames.QUICK_ADD_TASK,
    description: 'Quickly add one or more tasks using Todoist natural language input',
    parameters: ArgsSchema,

    async execute({ tasks }, client: TodoistApi) {
        const addTaskPromises = tasks.map((task) => processQuickAddTask(task, client))
        const newTasks = await Promise.all(addTaskPromises)
        const mappedTasks = newTasks.map(mapTask)

        return getToolOutput({
            textContent: generateTextContent(mappedTasks),
            structuredContent: {
                tasks: mappedTasks,
                totalCount: mappedTasks.length,
            },
        })
    },
} satisfies TodoistTool<typeof ArgsSchema>

async function processQuickAddTask(
    task: z.infer<typeof QuickAddTaskSchema>,
    client: TodoistApi,
): Promise<Task> {
    const apiArgs: QuickAddTaskArgs = { text: task.text }

    if (typeof task.note === 'string' && task.note.length > 0) apiArgs.note = task.note
    if (typeof task.reminder === 'string' && task.reminder.length > 0)
        apiArgs.reminder = task.reminder
    if (typeof task.autoReminder === 'boolean') apiArgs.autoReminder = task.autoReminder
    if (typeof task.meta === 'boolean') apiArgs.meta = task.meta

    return await client.quickAddTask(apiArgs)
}

function generateTextContent(tasks: ReturnType<typeof mapTask>[]) {
    const count = tasks.length
    const taskOrTasks = count === 1 ? 'task' : 'tasks'
    let summary = `Quick added ${count} ${taskOrTasks}.\n`
    summary += 'Tasks:\n'
    for (let i = 0; i < tasks.length; i++) {
        const t = tasks[i]!
        summary += `  - ${t.content}`
        if (t.dueDate) summary += ` | due: ${t.dueDate}`
        if (t.priority && t.priority !== 1) summary += ` | priority: P${t.priority}`
        summary += '\n'
    }
    return summary
}

export { quickAddTask }
