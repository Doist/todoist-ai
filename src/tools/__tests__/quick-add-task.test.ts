import type { Task, TodoistApi } from '@doist/todoist-api-typescript'
import { jest } from '@jest/globals'
import {
    createMockTask,
    extractStructuredContent,
    extractTextContent,
} from '../../utils/test-helpers.js'
import { ToolNames } from '../../utils/tool-names.js'
import { quickAddTask } from '../quick-add-task.js'

// Mock the Todoist API
const mockTodoistApi = {
    quickAddTask: jest.fn(),
} as unknown as jest.Mocked<TodoistApi>

const { QUICK_ADD_TASK } = ToolNames

describe(`${QUICK_ADD_TASK} tool`, () => {
    beforeEach(() => {
        jest.clearAllMocks()
    })

    describe('adding multiple tasks', () => {
        it('should quick add multiple tasks and return mapped results', async () => {
            // Mock API responses for each task
            const mockApiResponse1: Task = createMockTask({
                id: '9000000001',
                content: 'Quick task 1',
                url: 'https://todoist.com/showTask?id=9000000001',
            })
            const mockApiResponse2: Task = createMockTask({
                id: '9000000002',
                content: 'Quick task 2',
                description: 'Quick description',
                url: 'https://todoist.com/showTask?id=9000000002',
            })

            mockTodoistApi.quickAddTask
                .mockResolvedValueOnce(mockApiResponse1)
                .mockResolvedValueOnce(mockApiResponse2)

            const result = await quickAddTask.execute(
                {
                    tasks: [
                        { text: 'Quick task 1' },
                        { text: 'Quick task 2', note: 'Quick description' },
                    ],
                },
                mockTodoistApi,
            )

            // Verify API was called correctly for each task
            expect(mockTodoistApi.quickAddTask).toHaveBeenCalledTimes(2)
            expect(mockTodoistApi.quickAddTask).toHaveBeenNthCalledWith(1, {
                text: 'Quick task 1',
            })
            expect(mockTodoistApi.quickAddTask).toHaveBeenNthCalledWith(2, {
                text: 'Quick task 2',
                note: 'Quick description',
            })

            // Verify result is a concise summary
            expect(extractTextContent(result)).toMatchSnapshot()
            const structuredContent = extractStructuredContent(result)
            expect(Array.isArray((structuredContent as any).tasks)).toBe(true)
            expect((structuredContent as any).tasks).toHaveLength(2)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    totalCount: 2,
                    tasks: expect.arrayContaining([
                        expect.objectContaining({ id: '9000000001' }),
                        expect.objectContaining({ id: '9000000002' }),
                    ]),
                }),
            )
        })
    })

    describe('reminders and meta fields', () => {
        it('should handle quick add with reminder and meta', async () => {
            const mockApiResponse: Task = createMockTask({
                id: '9000000003',
                content: 'Quick task with reminder',
            })

            mockTodoistApi.quickAddTask.mockResolvedValue(mockApiResponse)

            const result = await quickAddTask.execute(
                {
                    tasks: [
                        {
                            text: 'Quick task with reminder',
                            reminder: 'tomorrow 9am',
                            meta: true,
                        },
                    ],
                },
                mockTodoistApi,
            )

            expect(mockTodoistApi.quickAddTask).toHaveBeenCalledWith({
                text: 'Quick task with reminder',
                reminder: 'tomorrow 9am',
                meta: true,
            })

            expect(extractTextContent(result)).toMatchSnapshot()
            const structuredContent = extractStructuredContent(result)
            expect(Array.isArray((structuredContent as any).tasks)).toBe(true)
            expect((structuredContent as any).tasks).toHaveLength(1)
            expect((structuredContent as any).tasks[0]).toEqual(
                expect.objectContaining({ id: '9000000003' }),
            )
        })
    })

    describe('error handling', () => {
        it('should propagate API errors', async () => {
            const apiError = new Error('API Error: Quick add failed')
            mockTodoistApi.quickAddTask.mockRejectedValue(apiError)

            await expect(
                quickAddTask.execute({ tasks: [{ text: 'fail' }] }, mockTodoistApi),
            ).rejects.toThrow(apiError.message)
        })
    })
})
