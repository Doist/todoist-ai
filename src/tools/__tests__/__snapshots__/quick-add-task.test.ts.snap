// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`quick-add-task tool should handle quick add with reminder and meta 1`] = `
"Quick added 1 task.
Tasks:
  - Quick task with reminder
"
`;

exports[`quick-add-task tool should quick add multiple tasks and return mapped results 1`] = `
"Quick added 2 tasks.
Tasks:
  - Quick task 1
  - Quick task 2
"
`;
